### PR TITLE
Add key to go back to previous autocomplete suggestion

### PR DESCRIPTION
**Pull request by kaustubh-nair**
Here's a gif 
reverse

### CODE REVIEW

**Comment by kaustubh-nair** [comment]
@rr- I've added `shift tab` as the default key to go back. Please have a look and let me know what you think, thanks!

File: urwid_readline/readline_edit.py [hunk 1]
```diff
@@ -372,7 +380,7 @@ def _complete(self):
         state = self._autocomplete_state
 
         match = self._autocomplete_func(state.infix, state.num)
-        if not match:
+        if (not match) or state.num < 0:
```
Comment (rr-) [inline urwid_readline/readline_edit.py]: Pointless quotes

File: urwid_readline/readline_edit.py [hunk 2]
```diff
@@ -338,18 +340,24 @@ def insert_new_line(self):
...
+    def _complete(self, cycle_forward):
         if self._autocomplete_state:
-            self._autocomplete_state.num += 1
+            if cycle_forward:
+                self._autocomplete_state.num += 1
+            else:
+                self._autocomplete_state.num -= 1
         else:
+            if not cycle_forward:
+                return
```
Comment (rr-) [inline urwid_readline/readline_edit.py]: Don't we want a wraparound behavior for autocomplete?

File: urwid_readline/readline_edit.py [hunk 3]
```diff
@@ -338,18 +340,24 @@ def insert_new_line(self):
         if self.multiline:
             self.insert_text("\n")
 
-    def enable_autocomplete(self, func, key=None):
+    def enable_autocomplete(self, func, key=None, key_reverse=None):
         self._autocomplete_func = func
         self._autocomplete_key = key if key is not None else "tab"
+        self._autocomplete_key_reverse = key_reverse if key_reverse is not None else "shift tab"
```
Comment (rr-) [inline urwid_readline/readline_edit.py]: Too long line, please pass the sources through black

```
python3 -m pip install --user black
python3 -m black -l79 .
```

File: urwid_readline/readline_edit.py [hunk 4]
```diff
@@ -120,8 +121,9 @@ def __init__(
 
     def keypress(self, size, key):
         self.size = size
-        if key == self._autocomplete_key and self._autocomplete_func:
-            self._complete()
+        if (key in {self._autocomplete_key, self._autocomplete_key_reverse}) and self._autocomplete_func:
+            cycle_forward = (key == self._autocomplete_key)
+            self._complete(cycle_forward)
```
Comment (rr-) [inline urwid_readline/readline_edit.py]:
```python3
        if key == self._autocomplete_key and self._autocomplete_func:
            self._complete(True)
            return None
        elif key == self._autocomplete_key_reverse and self._autocomplete_func:
            self._complete(False)
            return None
        else:
```

File: urwid_readline/readline_edit.py [hunk 5]
```diff
@@ -338,18 +340,24 @@ def insert_new_line(self):
...
         self._autocomplete_delims = delimiters
 
-    def _complete(self):
+    def _complete(self, cycle_forward):
         if self._autocomplete_state:
-            self._autocomplete_state.num += 1
+            if cycle_forward:
+                self._autocomplete_state.num += 1
+            else:
+                self._autocomplete_state.num -= 1
```
Comment (rr-) [inline urwid_readline/readline_edit.py]:
```diff
-            if cycle_forward:
-                self._autocomplete_state.num += 1
-            else:
-                self._autocomplete_state.num -= 1
+            self._autocomplete_state.num += 1 if cycle_forward else -1
```

File: urwid_readline/test_readline_edit.py [hunk 6]
```diff
@@ -470,6 +470,76 @@ def compl(text, state):
...
+
+        (
+            "",
+             0,
+             [("forward", "start", 5), ("forward", "stop", 4), ("forward", "next", 4), ("reverse", "stop", 4)]
+        ),
+        (
+            "",
+            0,
+            [("forward", "start", 5), ("forward", "stop", 4), ("forward", "next", 4), ("forward", "", 0), ("forward", "start", 5)],
```
Comment (rr-) [inline urwid_readline/test_readline_edit.py]: Too long lines, passing through black will take care of this

File: urwid_readline/test_readline_edit.py [hunk 7]
```diff
@@ -470,6 +470,76 @@ def compl(text, state):
...
+@pytest.mark.parametrize(
+    "autocomplete_key, autocomplete_key_reverse",
+    [
+        (None, None),
+        ("tab", "shift tab"),
+        ("ctrl q", "ctrl m"),
+    ]
+)
+def test_enable_autocomplete_reverse(
+         start_text, start_pos, edited, autocomplete_key, autocomplete_key_reverse
```
Comment (rr-) [inline urwid_readline/test_readline_edit.py]: Wrong indentation, let's pass through black

File: urwid_readline/readline_edit.py (hunk 3 above)
Comment (kaustubh-nair) [reply urwid_readline/readline_edit.py]: What version of black am I supposed to be using? 
I can't use the 179 flags for some reason
I get `Error: no such option: -1`

Comment (rr-) [reply urwid_readline/readline_edit.py]: Any version will be okay but the switch is `-l` not `-1`, as in `--line-length`

**Comment by kaustubh-nair** [comment]
@rr-  I've added wraparound for the autocomplete, but I had to take care of some cases manually. I  added a comment explaining why.
Also, black is not working for me with those flags, I'm not sure why

File: urwid_readline/readline_edit.py [hunk 8]
```diff
@@ -338,17 +342,23 @@ def insert_new_line(self):
...
         self._autocomplete_delims = delimiters
 
-    def _complete(self):
+    def _complete(self, cycle_forward):
         if self._autocomplete_state:
-            self._autocomplete_state.num += 1
+            if self._autocomplete_state.num == 0 and not cycle_forward:
+                self._autocomplete_state.num = None
+            elif self._autocomplete_state.num == -1 and cycle_forward:
+                self._autocomplete_state.num = None
```
Comment (kaustubh-nair) [inline urwid_readline/readline_edit.py]: Taking care of a couple of cases manually:
When you are at state 0 and cycle backwards, you would ideally want to get the original/non-autocompleted text. Setting state to -1 would set it to the last suggestion.
Same for when you're at -1 and cycle forward

Comment (kaustubh-nair) [reply urwid_readline/readline_edit.py]: Returning None causes a `TypeError` while indexing, so I've taken care of it there. I'm not entirely sure if this is the best way to handle the exception, so let me know if it's okay.

**Comment by kaustubh-nair** [comment]
@rr- Passed it through black :+1:

**Comment by rr-** [comment]
Thanks, released as 0.11

**Pull request merged**